### DISCUSSION TITLE
field BT-5141-Procedure 1.8

### DISCUSSION THREAD

**Discussion by elerik** [OP-TED/eForms-SDK#604] [Q&A] [answered]
Hi Labs 

this ID field is mandatory. 
But even it is not filled CVS is quite happy with it. 
From a group in where this field is i guess it misses condition ? like other fields in this groups ${(BT-727-Procedure is not present)} 
Can you please check it ? 

There is a whole group of fields at screen.
![image](https://github.com/OP-TED/eForms-SDK/assets/11737559/adc48c32-0054-4b34-8f9a-c62570616d47)

**Comment by YvesJo** [comment, marked as answer]
Hi,
For BT-5141-Procedure, there is a condition on the Forbidden rule and this condition is shared (in its negative form) with the mandatory rule. To be mandatory, the field must first be allowed (i.e. not forbidden). 

In the schematron:

```
<rule context="/*/cac:ProcurementProject/cac:RealizedLocation/cac:Address[$noticeSubType = '36']">

<!-- mandatory and forbidden rules -->
		<assert id="BR-BT-05141-0045" role="ERROR" diagnostics="BT-5141-Procedure" test="count(cac:Country/cbc:IdentificationCode) &gt; 0 or (cbc:Region/normalize-space(text()) = ('anyw','anyw-eea'))">rule|text|BR-BT-05141-0045</assert>
		<assert id="BR-BT-05141-0222" role="ERROR" diagnostics="BT-5141-Procedure" test="count(cac:Country/cbc:IdentificationCode) = 0 or not(cbc:Region/normalize-space(text()) = ('anyw','anyw-eea'))">rule|text|BR-BT-05141-0222</assert>
	</rule>
```
	
In fields.json:
please note that since the condition for the forbidden and mandatory are the same (one being the exact negation of the other), the condition is not reported in the mandatory one as the field must first be allowed

```
 {
    "id" : "BT-5141-Procedure",
    "parentNodeId" : "ND-ProcedurePlacePerformance",
    "name" : "Place Performance Country Code",
    "btId" : "BT-5141",
    "xpathAbsolute" : "/*/cac:ProcurementProject/cac:RealizedLocation/cac:Address/cac:Country/cbc:IdentificationCode",
    "xpathRelative" : "cac:Country/cbc:IdentificationCode",
    "xsdSequenceOrder" : [ { "cac:Country" : 28 }, { "cbc:IdentificationCode" : 2 } ],
    "type" : "code",
    "legalType" : "CODE",
    "repeatable" : {
      "value" : false,
      "severity" : "ERROR"
    },
    "forbidden" : {
      "value" : false,
      "severity" : "ERROR",
      "constraints" : [ {
...
      }, {
        "noticeTypes" : [ "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12", "13", "14", "15", "16", "17", "18", "19", "20", "21", "22", "23", "24", "29", "30", "31", "32", "33", "34", "35", "36", "37", "CEI" ],
        "condition" : "{ND-ProcedurePlacePerformance} ${BT-727-Procedure in ('anyw', 'anyw-eea')}",
        "value" : true,
        "severity" : "ERROR"
      } ]
    },
    "mandatory" : {
      "value" : false,
      "severity" : "ERROR",
      "constraints" : [ {
        "noticeTypes" : [ "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12", "13", "14", "15", "16", "17", "18", "19", "20", "21", "22", "23", "24", "29", "30", "31", "32", "33", "34", "35", "36", "37", "CEI" ],
        "value" : true,
        "severity" : "ERROR"
      } ]
    },
```

**Reply by elerik** [reply]
Hi, 
iam stuck on those conditions. I am going to share my thoughts now (from user perspective my testers are sometimes little bit crazy from results and very confused even now with some proper translations) 

dont know what is a best or at least optimal way to evaluate them. 
For example when Form is generated and all fields are empty. then BT-727-Procedure has no values. So its technically Mandatory. 
But then user start filling form. Select 'anyw' and then it should be forbidden. But field which was changed was BT-727-Procedure. 
From user perspective it should not appear anymore. 
I can solve it by request to "evaluate" whole form if everything is ok. And return some response. This is what i have now. But my users are confused. They want some "real time evaluation" which is challenge to obtain.
So iam thinking about some javascripts. But  then i have to run "validation" if any field is changed. Thats lots of operations. 
Other option is any field which is not directly mandatory simply dont request as required from user. 
Then they are not marked as required. And wait until user check if form data are valid. And then follow responses. Which will say for example "Hey this field is required because you have this selected" User find filed but that field is not marked as required. Hm Solution could be that in that moment whole form is refreshed with response in mind. Hmm. And adjust "required" fields. But it will every time require from user to "check if form is valid".
Do you have any feedback ideas how to optimize this process ? At this moment I evaluate only basic conditions like is present, is not present, == >= < and so  on (in C#) so iam trying to use CVS as much as possible.

**Reply by YvesJo** [reply]
indeed, having a dynamic UI with real time validation is a real challenge not only because of the technology, but also because of the business inherent complexity. On the NTD side, some actions are initiated to improve the user experience by reordering some fields so that they follow the logic. CVS is also used by some implementers and the error messages displayed back to the user in the appropriate context (e.g. next to the faulty field). It would be good for implementers to share their experience and approaches here, so I add some labels to this discussion for that.
